File: itmo-2s-golang-hw3/pkg/range/closed_range.go
package rangeI

import (
	"strconv"
)

type ClosedRange struct {
	min int
	max int
}

func (r ClosedRange) Length() int {
	return r.max - r.min + 1
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func (r ClosedRange) Intersect(other RangeInt) RangeInt {
	if other.IsEmpty() {
		return EmptyRange{}
	} else {
		otherMinimum, ok := other.Minimum()
		if !ok {
			panic("other min")
		}

		otherMaximum, ok := other.Maximum()
		if !ok {
			panic("other max")
		}

		intersectMin := max(r.min, otherMinimum)
		intersectMax := min(r.max, otherMaximum)

		if intersectMin <= intersectMax {
			return ClosedRange{intersectMin, intersectMax}
		} else {
			return EmptyRange{}
		}
	}
}

func (r ClosedRange) Union(other RangeInt) (RangeInt, bool) {
	if other.IsEmpty() {
		return r, true
	} else if r.IsIntersect(other) {

		otherMinimum, ok := other.Minimum()
		if !ok {
			panic("other min")
		}

		otherMaximum, ok := other.Maximum()
		if !ok {
			panic("other max")
		}

		unionMin := min(r.min, otherMinimum)
		unionMax := max(r.max, otherMaximum)
		return ClosedRange{unionMin, unionMax}, true
	} else {
		return nil, false
	}
}

func (r ClosedRange) IsEmpty() bool {
	return false
}

func (r ClosedRange) ContainsInt(i int) bool {
	return i >= r.min && i <= r.max
}

func (r ClosedRange) ContainsRange(other RangeInt) bool {
	if other.IsEmpty() {
		return true
	} else {
		otherMinimum, ok := other.Minimum()
		if !ok {
			panic("other min")
		}

		otherMaximum, ok := other.Maximum()
		if !ok {
			panic("other max")
		}

		return otherMinimum >= r.min && otherMaximum <= r.max
	}
}

func (r ClosedRange) IsIntersect(other RangeInt) bool {
	if other.IsEmpty() {
		return false
	} else {
		otherMinimum, ok := other.Minimum()
		if !ok {
			panic("other min")
		}

		otherMaximum, ok := other.Maximum()
		if !ok {
			panic("other max")
		}

		return r.min <= otherMaximum && r.max >= otherMinimum
	}
}

func (r ClosedRange) ToSlice() []int {
	a := make([]int, r.max-r.min+1)
	for i := range a {
		a[i] = r.min + i
	}
	return a
}

func (r ClosedRange) Minimum() (int, bool) {
	return r.min, true
}

func (r ClosedRange) Maximum() (int, bool) {
	return r.max, true
}

func (r ClosedRange) ToString() string {
	return "[" + strconv.Itoa(r.min) + ", " + strconv.Itoa(r.max) + "]"
}

File: itmo-2s-scala-hw3/src/main/scala/range/ClosedRange.scala
package range

class ClosedRange(val min: Int, val max: Int) extends Range {

  // Размер интервала
  override def length: Int = max - min + 1

  // Пересечение интервалов
  override def intersect(other: Range): Range = {
    if (other.isEmpty) EmptyRange
    else {
      val intersectMin = Math.max(min, other.minimum.get)
      val intersectMax = Math.min(max, other.maximum.get)
      if (intersectMin <= intersectMax) new ClosedRange(intersectMin, intersectMax)
      else Range.empty
    }
  }

  // Объединение пересекающихся интервалов
  override def union(other: Range): Option[Range] = {
    if (other.isEmpty) {
      Some(this)
    } else if (isIntersect(other)) {
      val unionMin = Math.min(min, other.minimum.get)
      val unionMax = Math.max(max, other.maximum.get)
      Some(new ClosedRange(unionMin, unionMax))
    } else None
  }

  // Проверка интервала на пустоту
  override def isEmpty: Boolean = false

  // Проверка точки на вхождение в интервал
  override def contains(i: Int): Boolean = i >= min && i <= max

  // Проверка интервала на вхождение в другой
  override def contains(other: Range): Boolean = {
    if (other.isEmpty) true
    else other.minimum.get >= min && other.maximum.get <= max
  }

  // Проверка двух интервалов на пересечение
  override def isIntersect(other: Range): Boolean = {
    if (other.isEmpty) false
    else min <= other.maximum.get && max >= other.minimum.get
  }

  // Перечисление всех точек интервала в порядке возрастания
  override def toList: List[Int] = List.range(min, max + 1)

  // Вычисление минимума интервала
  override def minimum: Option[Int] = Some(min)

  // Вычисление максимума интервала
  override def maximum: Option[Int] = Some(max)

  // Вычисление строкового представления интервала в формате [-3; 7], для пустого используйте ∅
  override def toString: String = s"[$min; $max]"

  // Проверка двух интервалов на равенство
  override def equals(obj: Any): Boolean = obj match {
    case r: ClosedRange => min == r.min && max == r.max
    case _              => false
  }

}
